symengine/matrices/conjugate_matrix.cpp

```diff
@@ -0,0 +1,139 @@
+    if (is_a<IdentityMatrix>(*arg) || is_a<ZeroMatrix>(*arg)
+        || is_a<DiagonalMatrix>(*arg) || is_a<ImmutableDenseMatrix>(*arg)
+        || is_a<ConjugateMatrix>(*arg) || is_a<Transpose>(*arg)
+        || is_a<MatrixAdd>(*arg) || is_a<HadamardProduct>(*arg)) {
+        return false;
+    }
+    return true;
+}
+
+int ConjugateMatrix::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<ConjugateMatrix>(o));
+
+    return arg_->compare(*down_cast<const ConjugateMatrix &>(o).arg_);
+}
+
+vec_basic ConjugateMatrix::get_args() const
+{
+    return {arg_};
+}
+
+class ConjugateMatrixVisitor : public BaseVisitor<ConjugateMatrixVisitor>
+{
+private:
+    RCP<const MatrixExpr> conjugate_;
+
+public:
+    ConjugateMatrixVisitor() {}
+
+    void bvisit(const Basic &x){};
+
+    void bvisit(const MatrixExpr &x)
+    {
+        auto arg = rcp_static_cast<const MatrixExpr>(x.rcp_from_this());
+        conjugate_ = make_rcp<const ConjugateMatrix>(arg);
+    }
+
+    void bvisit(const IdentityMatrix &x)
+    {
+        conjugate_ = rcp_static_cast<const MatrixExpr>(x.rcp_from_this());
+    }
+
+    void bvisit(const ZeroMatrix &x)
+    {
+        conjugate_ = rcp_static_cast<const MatrixExpr>(x.rcp_from_this());
+    }
+
+    void bvisit(const DiagonalMatrix &x)
+    {
+        auto diag = x.get_container();
+        vec_basic conj(diag.size());
+        for (size_t i = 0; i < diag.size(); i++) {
+            conj[i] = conjugate(diag[i]);
+        }
+        conjugate_ = make_rcp<const DiagonalMatrix>(conj);
+    }
+
+    void bvisit(const ImmutableDenseMatrix &x)
+    {
+        auto values = x.get_values();
+        vec_basic conj(values.size());
+        for (size_t i = 0; i < values.size(); i++) {
+            conj[i] = conjugate(values[i]);
+        }
+        conjugate_
+            = make_rcp<const ImmutableDenseMatrix>(x.nrows(), x.ncols(), conj);
+    }
+
+    void bvisit(const ConjugateMatrix &x)
+    {
+        conjugate_ = x.get_arg();
+    }
+
+    void bvisit(const Transpose &x)
+    {
+        // Shift order to transpose(conj(A))
+        auto arg = x.get_arg();
+        auto conj = make_rcp<const ConjugateMatrix>(arg);
+        conjugate_ = make_rcp<const Transpose>(conj);
+    }
+
+    void bvisit(const MatrixAdd &x)
+    {
+        vec_basic conj;
+        for (auto &e : x.get_terms()) {
+            e->accept(*this);
+            conj.push_back(conjugate_);
+        }
+        conjugate_ = make_rcp<const MatrixAdd>(conj);
+    }
+
+    void bvisit(const HadamardProduct &x)
+    {
+        vec_basic conj;
+        for (auto &e : x.get_factors()) {
+            e->accept(*this);
+            conj.push_back(conjugate_);
+        }
+        conjugate_ = make_rcp<const HadamardProduct>(conj);
+    }
+
+    RCP<const MatrixExpr> apply(const MatrixExpr &s)
+    {
+        s.accept(*this);
+        return conjugate_;
+    }
+};
+
+RCP<const MatrixExpr> conjugate_matrix(const RCP<const MatrixExpr> &arg)
```

Comment:
Why does this need a method that is separate from `conjugate`?

Reply:
Do you mean if we should overload this using the same name `conjugate` or if it should be implemented in the same function? We could use the same name, but I don't think this should be implemented in the same function. That would  create coupling between numbers and matrices and slow down the operation on numbers. Keeping them separate also makes it possible to (in some future) make the matrix expression module optional.

Reply:
what do you mean by numbers here? scalars or Number objects?

Reply:
I mean scalars.

Reply:
Then why have `MatrixExpr` be derived from `Basic`?

Reply:
Good question. Same reason as `Set` I guess. Perhaps they doesn't. Or it is good to root everything in Basic to allow common operations like `str` method and serialization for example. I envision a hierarchy like:

```
          Scalar  -> Add, Mul, Integer etc
Basic ->  Set
          Matrix  -> MatrixAdd, IdentityMatrix etc
```

Note that we don't have `Scalar` today. We could also in the future have other objects like abstract algebras. I think it is best to keep all of these things separate to not couple one type of object to another.

But if everything needs to be based on `Basic` I am not sure about. This is a very fundamental design question.